### ISSUE TITLE
sql/opt: zigzag join related planning cost is excessive

### ISSUE THREAD

**Issue by nvanbenschoten** [cockroachdb/cockroach#66226]
I've been digging into a large scale deployment of TPC-E that is saturating compute on a 96-node, 2880 vCPU cluster and noticed that a large amount of time (about 24% of all CPU) is spent in SQL planning under `(*Optimizer).Optimize`. This was surprising and seems excessive, given that I'm using prepared statements, but I do understand that there is still a substantial amount of work to be done once placeholders are filled in.

However, I think I have found a pattern that may explain a good chunk of this. Out of the 24% of time spent under `(*Optimizer).Optimize`, over 40% is spent in `(*CustomFuncs).GenerateZigzagJoins`. Put another way, over **10%** of all CPU time spent in this cluster is going to planning zigzag joins, even though the workload a) has no use for zigzag joins and b) uses join hints in almost all places to try to guide planning.

<img width="1659" alt="Screen Shot 2021-06-09 at 12 33 59 AM" src="https://user-images.githubusercontent.com/5438456/121294566-9c339d80-c8bb-11eb-9516-99252acf01aa.png">

[profile.zip](https://github.com/cockroachdb/cockroach/files/6621162/profile.zip)

Is something going wrong here? This seems quite excessive. And worse, we still seem to hit this cost even when the `sql.defaults.zigzag_join.enabled` cluster setting is set to false.

**Comment by nvanbenschoten** [comment]
For what it's worth, I believe a sizable chunk of the other time is `(*Optimizer).Optimize` is because the performance of `FastIntSet` degrades significantly once it hits the allocation-path, and we seem to use these all over the place in the optimizer with the assumption that we're hitting the fast-path, both to store state and to perform intermediate computation (like calls to `Union` and `Intersection` that throw away their result immediately). So once a query exceeds some number of columns, it has a much rougher time going through the optimizer.

As a result, `FastIntSet` is responsible for **1/3 of all memory allocations across the cluster** while running this intensive workload:
```
      File: cockroach
Type: alloc_objects
Time: Jun 9, 2021 at 4:58am (UTC)
Duration: 6.31s, Total samples = 83352720 
Active filters:
   focus=FastIntSet
Showing nodes accounting for 27478145, 32.97% of 83352720 total
----------------------------------------------------------+-------------
      flat  flat%   sum%        cum   cum%   calls calls% + context 	 	 
----------------------------------------------------------+-------------
                                           9831008 70.59% |   github.com/cockroachdb/cockroach/pkg/util.(*FastIntSet).Add /go/src/github.com/cockroachdb/cockroach/pkg/util/fast_int_set.go:75
                                           3342533 24.00% |   github.com/cockroachdb/cockroach/pkg/util.(*FastIntSet).UnionWith /go/src/github.com/cockroachdb/cockroach/pkg/util/fast_int_set.go:221
                                            753709  5.41% |   github.com/cockroachdb/cockroach/pkg/util.(*FastIntSet).DifferenceWith /go/src/github.com/cockroachdb/cockroach/pkg/util/fast_int_set.go:278
  13927250 16.71% 16.71%   13927250 16.71%                | github.com/cockroachdb/cockroach/pkg/util.(*FastIntSet).toLarge /go/src/github.com/cockroachdb/cockroach/pkg/util/fast_int_set.go:50
----------------------------------------------------------+-------------
                                           6136184 56.57% |   github.com/cockroachdb/cockroach/pkg/sql/opt.ColSet.Copy /go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/colset.go:58
                                           1998976 18.43% |   github.com/cockroachdb/cockroach/pkg/util.FastIntSet.Intersection /go/src/github.com/cockroachdb/cockroach/pkg/util/fast_int_set.go:255
                                           1482841 13.67% |   github.com/cockroachdb/cockroach/pkg/util.FastIntSet.Union /go/src/github.com/cockroachdb/cockroach/pkg/util/fast_int_set.go:234
                                           1228875 11.33% |   github.com/cockroachdb/cockroach/pkg/util.FastIntSet.Difference /go/src/github.com/cockroachdb/cockroach/pkg/util/fast_int_set.go:283
  10846876 13.01% 29.72%   10846876 13.01%                | github.com/cockroachdb/cockroach/pkg/util.FastIntSet.Copy /go/src/github.com/cockroachdb/cockroach/pkg/util/fast_int_set.go:190
----------------------------------------------------------+-------------
                                           9798237 99.67% |   github.com/cockroachdb/cockroach/pkg/sql/opt.(*ColSet).Add /go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/colset.go:33
                                             24578  0.25% |   github.com/cockroachdb/cockroach/pkg/sql/opt/xform.(*exploreState).markMemberAsFullyExplored /go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/xform/explorer.go:254
                                              8193 0.083% |   github.com/cockroachdb/cockroach/pkg/sql/opt/xform.(*groupState).markMemberAsFullyOptimized /go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:883
         0     0% 29.72%    9831008 11.79%                | github.com/cockroachdb/cockroach/pkg/util.(*FastIntSet).Add /go/src/github.com/cockroachdb/cockroach/pkg/util/fast_int_set.go:75
                                           9831008   100% |   github.com/cockroachdb/cockroach/pkg/util.(*FastIntSet).toLarge /go/src/github.com/cockroachdb/cockroach/pkg/util/fast_int_set.go:50
----------------------------------------------------------+-------------
```
Are we aware of this problem? Do we run microbenchmarks with the fast-path for `FastIntSet` disabled? Have we considered any kind of memory pooling for these `intsets.Sparse` objects?

**Comment by mgartner** [comment]
I think I should be able to speed up `GenerateZigZagJoins`. The majority of the time looks to be spent in `opt.(*TableMeta).IndexColumns`. There's already a [TODO](https://github.com/cockroachdb/cockroach/blob/7bb97675ba1a4e45cdaa5924ff74c5356da5d1ab/pkg/sql/opt/table_meta.go#L180) that hints that caching the result might be beneficial.

> And worse, we still seem to hit this cost even when the sql.defaults.zigzag_join.enabled cluster setting is set to false.

Are you sure about this? We check this session setting in a pretty straightforward way before doing any work in `Generate[InvertedIndex]ZigZagJoins`:

https://github.com/cockroachdb/cockroach/blob/7bb97675ba1a4e45cdaa5924ff74c5356da5d1ab/pkg/sql/opt/xform/select_funcs.go#L912-L915

https://github.com/cockroachdb/cockroach/blob/7bb97675ba1a4e45cdaa5924ff74c5356da5d1ab/pkg/sql/opt/xform/select_funcs.go#L1267-L1270

The function below `GenerateZigZagJoins` in that flame graph, `xform.(*scanIndexIter).ForEach`, is not called when the setting is disable. 

> Are we aware of this problem? Do we run microbenchmarks with the fast-path for FastIntSet disabled? Have we considered any kind of memory pooling for these intsets.Sparse objects?

I think we're aware of the issue, but I'm not sure what mitigations we've considered. I think this probably deserves an independent issue to track.

**Comment by nvanbenschoten** [comment]
> Are you sure about this? We check this session setting in a pretty straightforward way before doing any work in Generate[InvertedIndex]ZigZagJoins:

Oh interesting, we pull this from the session data, not from the cluster setting directly. So it's possible that I never created a new session after changing the cluster setting. Let me try again. Thanks for the tip!

**Comment by RaduBerinde** [comment]
> Oh interesting, we pull this from the session data, not from the cluster setting directly. 

Right, that is the case for all "sql.defaults.*" settings - they are defaults for session settings.

**Comment by RaduBerinde** [comment]
> uses join hints in almost all places

Zig-zag joins are used to implement certain filters, they aren't really a join algorithm.

I bet we could come up with some heuristics for cases where they surely won't help.

For the FastIntSet issue, do we have an idea for how many column IDs are in these plans? I'm a little surprised, because the plans we have for tpce don't use more than 64 columns (though the test catalog is missing at least the `tableoid` system column). Upping the support to 128 is an option.

**Comment by mgartner** [comment]
@nvanbenschoten I'm trying to write a small benchmark that reproduces a flamegraph similar to what you shared, but I'm having trouble. I can't reproduce a case where so much time is spent in `opt.(*TableMeta).IndexColumns`. Any information about the query or queries and the schema might be helpful, if you can provide it.

**Comment by RaduBerinde** [comment]
In any case, we should pre-calculate all the information in `optIndex` rather than doing any work in `Columns()`.

**Comment by RaduBerinde** [comment]
The lookup in that flamegraph is in FastIntMap, which has a similar problem to FastIntSet - beyond a certain number of columns, it falls back to a `map`. Make sure that in your benchmark, there are enough columns for that to happen.

**Comment by nvanbenschoten** [comment]
I don't know exactly which queries are exceeding the 64 column limit, but I suspect that they are queries with common-table expressions, which we use to run multi-statement txns as implicit transactions (a poor man's stored procedure). In general, most of the queries we use for TPC-E [have changed](https://github.com/cockroachdb/cockroach/issues/65728) since they were originally ported to our optimizer tests. We also have the duplicate index pattern applied to some of these tables 96-ways, but I don't think that would impact `Columns()`, would it?

**Comment by DrewKimball** [comment]
Maybe it would be beneficial to do for `FastIntMap` what was done for `FastIntSet` in #51635. Also, it might be a good idea to have the `large` set pick up where the `small` leaves off, instead of overlapping them. The current pattern is simple and good for read operations like intersects or contains, but not so much for things like union or intersection.

**Comment by RaduBerinde** [comment]
> We also have the duplicate index pattern applied to some of these tables 96-ways, but I don't think that would impact Columns(), would it?

If the zig-zag join rule is trying all pairs of indexes and we have 96 of them, that would explain why the zig-zag stuff is taking so long (it will still take up a visible fraction even after we fix the `Columns()` issue).

**Comment by mgartner** [comment]
Based on [this comment](https://github.com/cockroachdb/cockroach/blob/2c8e3d1913b9c23b7d94977e3c35b0e423e65de4/pkg/util/fast_int_map.go#L241-L242), 16 columns should be enough to make FastIntMap use the slow path.

**Comment by mgartner** [comment]
> We also have the duplicate index pattern applied to some of these tables 96-ways, but I don't think that would impact Columns(), would it?

Ok I think that explains it. I'm able to reproduce the issue if there are a HUGE number of indexes. In this duplicate index pattern, if there are 3 "base" indexes, each duplicated 96-ways, that's 288 indexes total on the table. Correct?

**Comment by nvanbenschoten** [comment]
> Ok I think that explains it. I'm able to reproduce the issue if there are a HUGE number of indexes. In this duplicate index pattern, if there are 3 "base" indexes, each duplicated 96-ways, that's 288 indexes total on the table. Correct?

Yes, that's correct. These are small tables, but they do have a very large number of indexes to provide local reads from all nodes without an RPC hop.

**Comment by mgartner** [comment]
@nvanbenschoten #66278 should reduce the planning costs in your workload.

**Comment by mgartner** [comment]
@nvanbenschoten I'm inclined to close this issue given that we've fixed the major issues with zigzag join planning. We can open a new one regarding `FastIntSet` performance if necessary. Does that sound good to you?

**Comment by nvanbenschoten** [comment]
That sounds good to me. Thanks again for the fix here @mgartner!
